File: lua/set.lua
function merge(a, b)
    local res = {}
    local i = 1
    local j = 1
    while i <= #a and j <= #b do
        if a[i] < b[j] then
            table.insert(res, a[i])
            i = i + 1
        elseif a[i] == b[j] then
            table.insert(res, a[i])
            table.insert(res, b[j])
            i = i + 1
            j = j + 1
        else
            table.insert(res, b[j])
            j = j + 1
        end
    end
    for k = j, #b do
        table.insert(res, b[k])
    end
    for k = i, #a do
        table.insert(res, a[k])
    end
    return res
end

function inplaceMerge(a, b)
    local i = 1
    local j = 1
    while i <= #a do
        while a[i] >= b[j] and j <= #b do
            table.insert(a, i, b[j])
            j = j + 1
        end
        i = i + 1
    end
    for k = j, #b do
        table.insert(a, b[k])
    end
    return a
end

-- a and b is sorted
function includes(a, b)
    if #a < #b then return false end
    local i = 1
    local j = 1
    while i <= #a and j <= #b do
        if a[i] == b[j] then
            i = i + 1
            j = j + 1
        elseif a[i] < b[j] then
            i = i + 1
        else
            j = j + 1
        end
    end
    return j == #b + 1
end

function union(a, b)
    local res = {}
    local i = 1
    local j = 1
    while i <= #a and j <= #b do
        if a[i] < b[j] then
            table.insert(res, a[i])
            i = i + 1
        elseif a[i] == b[j] then
            table.insert(res, a[i])
            i = i + 1
            j = j + 1
        else
            table.insert(res, b[j])
            j = j + 1
        end
    end
    for k = j, #b do
        table.insert(res, b[k])
    end
    for k = i, #a do
        table.insert(res, a[k])
    end
    return res
end

function difference(a, b)
    local res = {}
    local i = 1
    local j = 1
    while i <= #a and j <= #b do
        if a[i] < b[j] then
            table.insert(res, a[i])
            i = i + 1
        elseif a[i] == b[j] then
            i = i + 1
            j = j + 1
        else
            j = j + 1
        end
    end
    for k = i, #a do
        table.insert(res, a[k])
    end
    return res
end

function symmetricDifference(a, b)
    return union(difference(a, b), difference(b, a))
end

function intersection(a, b)
    return difference(union(a, b), symmetricDifference(a, b))
end

return {
    merge = merge,
    inplaceMerge = inplaceMerge,
    includes = includes,
    union = union,
    difference = difference,
    symmetricDifference = symmetricDifference,
    intersection = intersection
}

File: php/Set.php
<?php
namespace Algorithm;

function merge(array $a, array $b) {
    $res = [];
    $len_a = \count($a);
    $len_b = \count($b);
    $i = $j = 0;
    while ($i < $len_a && $j < $len_b) {
        $x = $a[$i];
        $y = $b[$j];
        if ($x < $y) {
            $res[] = $x;
            $i++;
        } elseif ($x > $y) {
            $res[] = $y;
            $j++;
        } else {
            $res[] = $x;
            $res[] = $y;
            $i++;
            $j++;
        }
    }
    for (; $i < $len_a; $i++)
        $res[] = $a[$i];
    for (; $j < $len_b; $j++)
        $res[] = $b[$j];
    return $res;
}

function inplaceMerge(array &$a, array $b) {
    $a = array_values(merge($a, $b));
}

function includes(array $ary, array $scope) {
    $end = \count($scope);
    if ($end == 0)
        return true;
    $j = 0;
    for ($i = 0; $i < \count($ary); $i++) {
        if ($ary[$i] == $scope[$j]) {
            $j++;
            if ($j == $end)
                return true;
        }
    }
    return false;
}

function union(array $a, array $b) {
    $res = [];
    $len_a = \count($a);
    $len_b = \count($b);
    $i = $j = 0;
    while ($i < $len_a && $j < $len_b) {
        $x = $a[$i];
        $y = $b[$j];
        if ($x < $y) {
            $res[] = $x;
            $i++;
        } elseif ($x > $y) {
            $res[] = $y;
            $j++;
        } else {
            $res[] = $x;
            $i++;
            $j++;
        }
    }
    for (; $i < $len_a; $i++)
        $res[] = $a[$i];
    for (; $j < $len_b; $j++)
        $res[] = $b[$j];
    return $res;
}

function difference(array $a, array $b) {
    $res = [];
    $len_a = \count($a);
    $len_b = \count($b);
    $i = $j = 0;
    while ($i < $len_a && $j < $len_b) {
        $x = $a[$i];
        $y = $b[$j];
        if ($x < $y) {
            $res[] = $x;
            $i++;
        } elseif ($x > $y) {
            $j++;
        } else {
            $i++;
            $j++;
        }
    }
    for (; $i < $len_a; $i++)
        $res[] = $a[$i];
    return $res;
}

function symmetricDifference(array $a, array $b) {
    $res = [];
    $len_a = \count($a);
    $len_b = \count($b);
    $i = $j = 0;
    while ($i < $len_a && $j < $len_b) {
        $x = $a[$i];
        $y = $b[$j];
        if ($x < $y) {
            $res[] = $x;
            $i++;
        } elseif ($x > $y) {
            $res[] = $y;
            $j++;
        } else {
            $i++;
            $j++;
        }
    }
    for (; $i < $len_a; $i++)
        $res[] = $a[$i];
    for (; $j < $len_b; $j++)
        $res[] = $b[$j];
    return $res;
}

function intersection(array $a, array $b) {
    $res = [];
    $len_a = \count($a);
    $len_b = \count($b);
    $i = $j = 0;
    while ($i < $len_a && $j < $len_b) {
        $x = $a[$i];
        $y = $b[$j];
        if ($x == $y) {
            $res[] = $x;
            $i++;
            $j++;
        } elseif ($x > $y) {
            $j++;
        } else {
            $i++;
        }
    }
    return $res;
}
?>
